File: cpp/src/factory.cpp
// factory.cpp
#include <iostream>
#include <string>
#include <memory>

// 추상 클래스
class Animal {
public:
    virtual std::string speak() = 0;
    virtual ~Animal() = default;
};

// 구체 클래스들
class Dog : public Animal {
public:
    std::string speak() override {
        return "Woof!";
    }
};

class Cat : public Animal {
public:
    std::string speak() override {
        return "Meow!";
    }
};

// 팩토리 클래스
class AnimalFactory {
public:
    std::unique_ptr<Animal> createAnimal(const std::string& animalType) {
        if (animalType == "dog") {
            return std::make_unique<Dog>();
        } else if (animalType == "cat") {
            return std::make_unique<Cat>();
        }
        return nullptr;
    }
};

// 테스트 코드
int main() {
    AnimalFactory factory;
    
    auto dog = factory.createAnimal("dog");
    std::cout << "Dog says: " << dog->speak() << std::endl;
    
    auto cat = factory.createAnimal("cat");
    std::cout << "Cat says: " << cat->speak() << std::endl;
    
    return 0;
}

File: kotlin/src/main/kotlin/Factory.kt
// Factory.kt
package designpatterns

// 추상 클래스
abstract class Animal {
    abstract fun speak(): String
}

// 구체 클래스들
class Dog : Animal() {
    override fun speak(): String {
        return "Woof!"
    }
}

class Cat : Animal() {
    override fun speak(): String {
        return "Meow!"
    }
}

// 팩토리 클래스
class AnimalFactory {
    fun createAnimal(animalType: String): Animal? {
        return when (animalType.lowercase()) {
            "dog" -> Dog()
            "cat" -> Cat()
            else -> null
        }
    }
}

// 테스트 함수
fun testFactory() {
    val factory = AnimalFactory()
    
    val dog = factory.createAnimal("dog")
    dog?.let { println("Dog says: ${it.speak()}") }
    
    val cat = factory.createAnimal("cat")
    cat?.let { println("Cat says: ${it.speak()}") }
}